# Машинное обучение. Лабораторная работа 1

## Настройка среды

In [1]:
%pip install numpy
%pip install pandas
%pip install sklearn

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


## Задача и датасет

Датасет взят с сайта : http://archive.ics.uci.edu/ml/datasets/Mushroom

В базе данных представлена информацию о видах грибов, их свойствах и особенностях. По описанию гриба при помощи искусственного интеллекта будем решать задачу классификации грибов на ядовитые и съедобные

## Датасет

In [2]:
import pandas as pds

df = pds.read_csv('agaricus-lepiota.data')

In [3]:
# df

## Описание атрибутов и возможных значений: 

0. edible : p=poisonous, e=edible
1. cap-shape: bell=b,conical=c,convex=x,flat=f, knobbed=k,sunken=s
2. cap-surface: fibrous=f,grooves=g,scaly=y,smooth=s
3. cap-color: brown=n,buff=b,cinnamon=c,gray=g,green=r, pink=p,purple=u,red=e,white=w,yellow=y
4. bruises?: bruises=t,no=f
5. odor: almond=a,anise=l,creosote=c,fishy=y,foul=f, musty=m,none=n,pungent=p,spicy=s
6. gill-attachment: attached=a,descending=d,free=f,notched=n
7. gill-spacing: close=c,crowded=w,distant=d
8. gill-size: broad=b,narrow=n
9. gill-color: black=k,brown=n,buff=b,chocolate=h,gray=g, green=r,orange=o,pink=p,purple=u,red=e, white=w,yellow=y
10. stalk-shape: enlarging=e,tapering=t
11. stalk-root: bulbous=b,club=c,cup=u,equal=e, rhizomorphs=z,rooted=r,missing=?
12. stalk-surface-above-ring: fibrous=f,scaly=y,silky=k,smooth=s
13. stalk-surface-below-ring: fibrous=f,scaly=y,silky=k,smooth=s
14. stalk-color-above-ring: brown=n,buff=b,cinnamon=c,gray=g,orange=o, pink=p,red=e,white=w,yellow=y
15. stalk-color-below-ring: brown=n,buff=b,cinnamon=c,gray=g,orange=o, pink=p,red=e,white=w,yellow=y
16. veil-type: partial=p,universal=u
17. veil-color: brown=n,orange=o,white=w,yellow=y
18. ring-number: none=n,one=o,two=t
19. ring-type: cobwebby=c,evanescent=e,flaring=f,large=l, none=n,pendant=p,sheathing=s,zone=z
20. spore-print-color: black=k,brown=n,buff=b,chocolate=h,green=r, orange=o,purple=u,white=w,yellow=y
21. population: abundant=a,clustered=c,numerous=n, scattered=s,several=v,solitary=y
22. habitat: grasses=g,leaves=l,meadows=m,paths=p, urban=u,waste=w,woods=d

Разделим датасет на обучающую и тестовую выборки. Так как записей в датасете - 8123, разделим датасет в соотношении 3:1 

In [4]:
separate_index = int(3.0*len(df)/4.0)
print(separate_index)

6093


In [5]:
df_learn = df.iloc[0:separate_index]
df_test = df.iloc[separate_index:]

# Алгоритм KNN

## Подготовка датасета

Реализуем алгоритм KNN, сначала сопоставим всем качественным признакам числовые значения и будем работать с числовыми значениями.

In [6]:
edible_dict = { 'p' : 0, 'e' : 1 }
cap_shape_dict = { 'b' : 0, 'c' : 1, 'x' : 2, 'f' : 3, 'k' : 4, 's' : 5 }
cap_surface_dict = { 'f' : 0, 'g' : 1, 'y' : 2, 's' : 3 }
cap_color_dict = { 'n' : 0, 'b' : 1, 'c' : 2, 'g' : 3, 'r' : 4, 'p' : 5, 'u' : 6, 'e' : 7, 'w' : 8, 'y' : 9 }
bruises_dict = { 't' : 0, 'f' : 1 }
odor_dict = { 'a' : 0, 'l' : 1, 'c' : 2, 'y' : 3, 'f' : 4, 'm' : 5, 'n' : 6, 'p' : 7, 's' : 8 }
gill_attachment_dict = { 'a' : 0, 'd' : 1, 'f' : 2, 'n' : 3 }
gill_spacing_dict = { 'c' : 0, 'w' : 1, 'd' : 2 }
gill_size_dict = { 'b' : 0, 'n' : 1 }
gill_color_dict = { 'k' : 0, 'n' : 1, 'b' : 2, 'h' : 3, 'g' : 4, 'r' : 5, 'o' : 6, 'p' : 7, 'u' : 8, 'e' : 9, 'w' : 10, 'y' : 11 }
stalk_shape_dict = { 'e' : 0, 't' : 1 }
stalk_root_dict = { 'b' : 0, 'c' : 1, 'u' : 2, 'e' : 3, 'z' : 4, 'r' : 5, '?' : 6 }
stalk_surface_above_ring_dict = { 'f' : 0, 'y' : 1, 'k' : 2, 's' : 3 }
stalk_surface_below_ring_dict = { 'f' : 0, 'y' : 1, 'k' : 2, 's' : 3 }
stalk_color_above_ring_dict = { 'n' : 0, 'b' : 1, 'c' : 2, 'g' : 3, 'o' : 4, 'p' : 5, 'e' : 6, 'w' : 7, 'y' : 8 }
stalk_color_below_ring_dict = { 'n' : 0, 'b' : 1, 'c' : 2, 'g' : 3, 'o' : 4, 'p' : 5, 'e' : 6, 'w' : 7, 'y' : 8 }
veil_type_dict = { 'p' : 0, 'u' : 1 }
veil_color_dict = { 'n' : 0, 'o' : 1, 'w' : 2, 'y' : 3 }
ring_number_dict = { 'n' : 0, 'o' : 1, 't' : 2 }
ring_type_dict = { 'c' : 0, 'e' : 1, 'f' : 2, 'l' : 3, 'n' : 4, 'p' : 5, 's' : 6, 'z' : 7 }
spore_print_color_dict = { 'k' : 0, 'n' : 1, 'b' : 2, 'h' : 3, 'r' : 4, 'o' : 5, 'u' : 6, 'w' : 7, 'y' : 8 }
population_dict = { 'a' : 0, 'c' : 1, 'n' : 2, 's' : 3, 'v' : 4, 'y' : 5 }
habitat_dict = { 'g' : 0, 'l' : 1, 'm' : 2, 'p' : 3, 'u' : 4, 'w' : 5, 'd' : 6 }

def replace_attribute_on_number(row):
    row['edible'] = edible_dict[row['edible']] # 1
    row['cap-shape'] = cap_shape_dict[row['cap-shape']] # 2
    row['cap-surface'] = cap_surface_dict[row['cap-surface']] # 3
    row['cap-color'] = cap_color_dict[row['cap-color']] # 4
    row['bruises'] = bruises_dict[row['bruises']] # 5
    row['odor'] = odor_dict[row['odor']] # 6
    row['gill-attachment'] = gill_attachment_dict[row['gill-attachment']] # 7
    row['gill-spacing'] = gill_spacing_dict[row['gill-spacing']] # 8
    row['gill-size'] = gill_size_dict[row['gill-size']] # 9
    row['gill-color'] = gill_color_dict[row['gill-color']] # 10
    row['stalk-shape'] = stalk_shape_dict[row['stalk-shape']] # 11
    row['stalk-root'] = stalk_root_dict[row['stalk-root']] # 12
    row['stalk-surface-above-ring'] = stalk_surface_above_ring_dict[row['stalk-surface-above-ring']] # 13
    row['stalk-surface-below-ring'] = stalk_surface_below_ring_dict[row['stalk-surface-below-ring']] # 14
    row['stalk-color-above-ring'] = stalk_color_above_ring_dict[row['stalk-color-above-ring']] # 15
    row['stalk-color-below-ring'] = stalk_color_below_ring_dict[row['stalk-color-below-ring']] # 16
    row['veil-type'] = veil_type_dict[row['veil-type']] # 17
    row['veil-color'] = veil_color_dict[row['veil-color']] # 18
    row['ring-number'] = ring_number_dict[row['ring-number']] # 19
    row['ring-type'] = ring_type_dict[row['ring-type']] # 20
    row['spore-print-color'] = spore_print_color_dict[row['spore-print-color']] # 21
    row['population'] = population_dict[row['population']] # 22
    row['habitat'] = habitat_dict[row['habitat']] # 23


def change_atr_to_numbers_in_df(df):
    df_copy = df.copy()
    for index, row in df_copy.iterrows():
        replace_attribute_on_number(row)
    return df_copy

df_learn_numeric = change_atr_to_numbers_in_df(df_learn)


In [7]:
# df_learn_numeric

Получим ядовитые и съедобные грибы

In [8]:
df_learn_poisonous = df_learn_numeric[df_learn_numeric['edible'] == 0]
df_learn_edible =    df_learn_numeric[df_learn_numeric['edible'] == 1]

Проверим, что разделение прошло успешно

In [9]:
print((len(df_learn_edible) + len(df_learn_poisonous) == len(df_learn_numeric)))

True


## Функционал

Датасет подготовлен для обучения модели. Введем несколько дополнительных функций: эвклидово расстояние по которому будем рассчитывать расстояние и функцию взятия среднего арифметического. 

In [10]:
from math import sqrt
def euclidean(v1, v2):
    return sqrt(sum([(x - y)**2 for x, y in zip(v1, v2)]))

def get_middle_point(df):
    result = []
    df_lists = df.values.tolist()
    for i in range(1, 23):
        sum = 0
        for llist in df_lists:
            sum += llist[i]
        result.append(sum/len(df))
    return result

## Обучение модели

Найдем две 22-мерные точки соотвествующие вредным и съедобным грибам

In [11]:
poisonous_point = []
edible_point = []

def learn_process_knn():
    edible_point = get_middle_point(df_learn_edible)
    poisonous_point = get_middle_point(df_learn_poisonous)
    return edible_point, poisonous_point

edible_point, poisonous_point = learn_process_knn()  
print("Центр группы съедобных грибов: {}".format(edible_point))
print()
print("Центр группы ядовитых грибов:  {}".format(poisonous_point))

Центр группы съедобных грибов: [2.3448275862068964, 1.4439315775183275, 4.2682595710019005, 0.26147162639152866, 4.805321748574531, 1.9989139288623405, 0.2478957371707847, 0.0781971219114852, 5.415965245723595, 0.7037740972033668, 1.4993212055389629, 2.667390714091773, 2.524029323920717, 6.03393972305186, 5.94270974748846, 0.0, 1.9989139288623405, 1.0524029323920716, 3.8661417322834644, 1.0363833831115938, 3.366549008960087, 3.645397773554168]

Центр группы ядовитых грибов:  [2.3991701244813277, 1.6157676348547718, 5.003734439834025, 0.7410788381742739, 4.34149377593361, 2.0, 0.043153526970954356, 0.3128630705394191, 4.75850622406639, 0.22987551867219916, 1.0605809128630705, 2.211203319502075, 2.198755186721992, 4.197510373443984, 4.154356846473029, 0.0, 2.0, 1.029875518672199, 3.429875518672199, 3.0730290456431537, 4.106224066390041, 3.0182572614107883]


Обучили модель, теперь проведем тестирование

### Тестирование

In [12]:
def is_edible_knn(row):
    len_to_poisonous = euclidean(row[1:], poisonous_point)
    len_to_edible = euclidean(row[1:], edible_point)
    return len_to_edible < len_to_poisonous
        
def test_knn(df):
    df_numeric = change_atr_to_numbers_in_df(df)
    success_count = 0
    for index, row in df_numeric.iterrows():
        real_answer = (row['edible'] == 1)
        algo_answer = is_edible_knn(row)
        if real_answer == algo_answer:
            success_count += 1
    return success_count/len(df)
    
print("Точность алгоритма: {}%".format(test_knn(df_test)*100))

Точность алгоритма: 83.89955686853766%


## Наивный байесовский классификатор

Реализуем простейший байесовский классификатор

Для начала посчитаем сколько в обучающей выборке встречаются ядовитые и съедобвные грибы

In [13]:
count_poisonous = len(df_learn_poisonous)
count_edible = len(df_learn_edible)
print("Ядовитых грибов: {}, съедобных: {}".format(count_poisonous,count_edible))

Ядовитых грибов: 2410, съедобных: 3683


Посчитаем вероятности встретить каждый признак в каждом классе

In [14]:
def get_dict_probabilities(df):
    res = list()
    df_lists = df.values.tolist()
    for i in range(1,23):
        atr_probabilities = dict()
        for llist in df_lists:
            if llist[i] in atr_probabilities:
                atr_probabilities[llist[i]] += 1
            else :
                atr_probabilities[llist[i]] = 1
        for item in atr_probabilities:
            atr_probabilities[item] /= len(df)
        res.append(atr_probabilities)
    return res

dict_poisonous = get_dict_probabilities(df_learn_poisonous)
dict_edible = get_dict_probabilities(df_learn_edible)

### Тестирование

In [15]:
def is_edible_bayes(row):
    list_row = row.tolist()
    sum_poisonous = 0
    sum_edible = 0
    for i in range(1,23):
        if list_row[i] in dict_poisonous[i-1]:
            sum_poisonous += dict_poisonous[i-1][list_row[i]]
        if list_row[i] in dict_edible[i-1]:
            sum_edible += dict_edible[i-1][list_row[i]]
    return sum_edible > sum_poisonous

def test_bayes(df):
    df_numeric = change_atr_to_numbers_in_df(df)
    success_count = 0
    for index, row in df_numeric.iterrows():
        real_answer = (row['edible'] == 1)
        algo_answer = is_edible_bayes(row)
        if real_answer == algo_answer:
            success_count += 1
    return success_count/len(df)

print("Точность алгоритма: {}%".format(test_bayes(df_test)*100))

Точность алгоритма: 83.45642540620383%


## Сравнение с реализацией sklearn

### Алгоритм KNN

Подготовим данные для алгоритма

In [16]:
df_sklearn_train_data = list()
df_sklearn_train_ans = list()
for index, row in df_learn_poisonous.iterrows():
    df_sklearn_train_data.append(row.values[1:])
    df_sklearn_train_ans.append(row.values[0])
for index, row in df_learn_edible.iterrows():
    df_sklearn_train_data.append(row.values[1:])
    df_sklearn_train_ans.append(row.values[0])
    
df_sklearn_test_data = list()
df_sklearn_test_ans = list()
for index, row in change_atr_to_numbers_in_df(df_test).iterrows():
    df_sklearn_test_data.append(row.values[1:])
    df_sklearn_test_ans.append(row.values[0])

Проводим тестирование

In [17]:
from sklearn.neighbors import KNeighborsClassifier

nbrs = KNeighborsClassifier(n_neighbors=10, algorithm='auto').fit(df_sklearn_train_data, df_sklearn_train_ans)

print("Точность алгоритма : {}%".format(nbrs.score(df_sklearn_test_data, df_sklearn_test_ans)*100))

Точность алгоритма : 81.33924175283111%


### Байесовский классификатор

Проводим тестирование

In [18]:
from sklearn.naive_bayes import GaussianNB

gnb = GaussianNB().fit(df_sklearn_train_data, df_sklearn_train_ans)

print("Точность алгоритма : {}%".format(gnb.score(df_sklearn_test_data, df_sklearn_test_ans)*100))

Точность алгоритма : 96.30723781388478%


## Выводы

В ходе лабораторной работы были реализованы простейшие классификаторы KNN и байесовкий с приблизительной точностью для выбранного мной датасета в 83% каждый. Сравнил собственную реализацию с реализациями в библиотеке sklearn. Точность KNN собственной реализации оказалась примерно такой же как и в sklearn, алгоритм байесовского классификатора в sklearn имеет удивительно высокую точность в 96%, точность байесовского классификатора сильно выше моей собственной реализации.

В ходе выполнения работы смог убедится на практике в том, что при помощи математики, вычислительной техники и данных можно находить различные корреляции между метриками, которые не поддаются обычному разъяснению для единичных сущностей, иначе говоря, на основе статистики можно с большой долей вероятности предсказать некоторые свойства выбранного объекта. 